### PR TITLE
Deployment to zkSync, Polygon zkevm

### PR DESCRIPTION
**Pull request by kevin-fruitful**
The motivation behind this PR is to be able to deploy to zkSync.

- [x] Create Diamond ABI to be used with hardhat/ethers - zkSync doesn't like this ABI generated
- [ ] https://era.zksync.io/docs/dev/developer-guides/aa.html - new docs already?!
- [ ] Remove usage of `extcodesize`
- [x] Script to deploy all facets, diamondCut() and initialize state
- [x] Mock call to uniswap v3 to fetch price - MockTwap contract can set sqrtPriceX96 when twap interval = 0
- [x] Refactor meme purchasing to accept base coin

### CODE REVIEW

File: script/DeployToZksync.s.sol [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: MIT
```
Comment (hiddentao) [inline script/DeployToZksync.s.sol:1]: Is this file still needed?

File: src/facets/MemeMarketFacet.sol [hunk 2]
```diff
@@ -27,34 +31,59 @@ contract MemeMarketFacet is FacetBase {
...
 
+        require(msg.value >= cost, "Not enough coin sent");
+
         // first, transfer wmatic to treasury
-        IERC20 wmatic = IERC20(s.currencyAddress);
-        wmatic.transferFrom(_msgSender(), s.addresses[LibConstants.TREASURY_ADDRESS], amount);
+        // IERC20 wmatic = IERC20(s.currencyAddress);
+        // wmatic.transferFrom(_msgSender(), s.addresses[LibConstants.TREASURY_ADDRESS], amount);
+
+        // send coin to treasury, which is the server address
```
Comment (hiddentao) [inline src/facets/MemeMarketFacet.sol:77]: Let's keep it as the treasury address constant for now, not to the server address directly. Our server address is a hot wallet and should only hold what's needed to transact, not our entire treasury. 

Instead, let's have a check function which, when called, checks if the server balance is below 1 eth and if so it then sends it some funds from the treasury. Let's call this new check function here at the bottom after receiving funds. Please make this check function public, so that anyone can call it at any time.

File: src/shared/ReentrancyGuard.sol [hunk 3]
```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: MIT
```
Comment (hiddentao) [inline src/shared/ReentrancyGuard.sol:1]: Why are we adding this to our code? why can't we just import it?

File: test/MemeTest.sol [hunk 4]
```diff
@@ -37,6 +38,36 @@ contract MemeTest is TestBaseContract {
         proxy.buyMeme(MemeBuySizeDollars.Five);
     }
 
+    function testZkSyncMockPurchaseMeme() public {
```
Comment (hiddentao) [inline test/MemeTest.sol:41]: Change to `testPurchaseMemeWithMockTwap`. And change the following method to `testPurchaseMemeWithLiveTwap`.

`testBuyMemeWithInsufficientWMatic` should be replaced with one which uses the mock to set a price that isn't then met when calling buy.

File: test/MemeTest.sol [hunk 5]
```diff
@@ -37,6 +38,36 @@ contract MemeTest is TestBaseContract {
...
+        uint256 receiverBalance = address(proxy.getAddress(LibConstants.SERVER_ADDRESS)).balance;
+
+        console2.log("receiver balance", receiverBalance);
+        console2.log("user1 ending balance", user1_endingBalance);
+        assertEq(
+            user1_startingBalance,
+            user1_endingBalance + receiverBalance,
+            "user1 balance is not correct"
+        );
+    }
```
Comment (hiddentao) [inline test/MemeTest.sol:69]: I can't see where this test is checking the amount of MEME received by the user? What's the point of the test if it isn't checking that?

**Review by hiddentao — CHANGES_REQUESTED** [review]
Please see comments.

File: src/shared/ReentrancyGuard.sol (hunk 3 above)
Comment (kevin-fruitful) [reply src/shared/ReentrancyGuard.sol:1]: It's refactored to be used with our AppStorage storage pattern. It's not recommended we have any storage variables local to a contract since we are delegatecalling into them and therefore confusing which storage slot we are writing into.

File: script/DeployToZksync.s.sol (hunk 1 above)
Comment (kevin-fruitful) [reply script/DeployToZksync.s.sol]: Removed for now, working foundry plugin coming soon hopefully :)

File: src/facets/MemeMarketFacet.sol (hunk 2 above)
Comment (kevin-fruitful) [reply src/facets/MemeMarketFacet.sol]: Created LibComptroller._routeCoinPayment() to route coin payments to server address if the server has less than 0.5 eth, routes to the proxy otherwise.

File: test/MemeTest.sol (hunk 5 above)
Comment (kevin-fruitful) [reply test/MemeTest.sol:76]: fixed, will make the tests for meme more robust after merging

File: test/MemeTest.sol (hunk 4 above)
Comment (kevin-fruitful) [reply test/MemeTest.sol:46]: fixed, and meme tests are now dealing with coin instead of erc20 tokens

File: src/libs/LibComptroller.sol [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        address serverAddress = s.addresses[LibConstants.SERVER_ADDRESS];
+        if (address(serverAddress).balance < 0.5 ether) {
+            // send coin to the server address
+            (bool sent, ) = address(serverAddress).call{ value: amount }("");
+            if (!sent) {
+                revert FailedToSendEtherToServer(serverAddress);
+            }
+        } else {
+            // send coin to the treasury address, which is the proxy / diamond contract
+            (bool sent, ) = address(this).call{ value: amount }("");
```
Comment (hiddentao) [inline src/libs/LibComptroller.sol:23]: Why do we need to do this? Hasn't `this` already received the ETH?

File: src/shared/ReentrancyGuard.sol [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    // slot's contents, replace the bits taken up by the boolean, and then write
+    // back. This is the compiler's defense against contract upgrades and
+    // pointer aliasing, and it cannot be disabled.
+
+    // The values being non-zero value makes deployment a bit more expensive,
+    // but in exchange the refund on every call to nonReentrant will be lower in
+    // amount. Since refunds are capped to a percentage of the total
+    // transaction's gas, it is best to keep them low in cases like this one, to
+    // increase the likelihood of the full refund coming into effect.
+    uint256 private constant _NOT_ENTERED = 1;
```
Comment (hiddentao) [inline src/shared/ReentrancyGuard.sol:36]: Can't we use an enum here?

Comment (kevin-fruitful) [reply src/shared/ReentrancyGuard.sol:36]: this is cheaper

File: test/MemeTest.sol (hunk 4 above)
Comment (hiddentao) [reply test/MemeTest.sol:46]: Please rename the method.

File: src/libs/LibComptroller.sol (hunk 6 above)
Comment (kevin-fruitful) [reply src/libs/LibComptroller.sol:23]: oh true, let me fix

File: test/MemeTest.sol [hunk 8]
```diff
@@ -28,41 +36,105 @@ contract MemeTest is TestBaseContract {
...
 
-        uint256 newBal = wmatic.balanceOf(address(this));
-        console2.log("new balance", newBal);
-        assertEq(memeToken.balanceOf(address(this)), 1000 ether);
+        console2.log(address(this).balance);
+
+        address user1 = address(0x1);
+
+        // give more than 0.5 coin to server
+        vm.deal(SERVER_ADDRESS, 1 ether);
```
Comment (hiddentao) [inline test/MemeTest.sol:92]: Please add a test which does the opposite and check that server gets the funds. Then check that once server balance is over 0.5 ETH it no longer gets funds. Also, let's make the 0.5 ETH a constant in LibConstants so that we don't hardcode it anywhere.

File: test/MemeTest.sol [hunk 9]
```diff
@@ -28,41 +36,105 @@ contract MemeTest is TestBaseContract {
...
-        uint256 balProxy = wmatic.balanceOf(address(proxy));
-        assertEq(newBal + balProxy, prevBal, "wmatic balances are not correct");
+        assertEq(memeToken.balanceOf(user1), 1000 ether);
 
         // todo add more checks
-        proxy.buyMeme(MemeBuySizeDollars.Ten);
-        proxy.buyMeme(MemeBuySizeDollars.Twenty);
-        proxy.buyMeme(MemeBuySizeDollars.Fifty);
-        proxy.buyMeme(MemeBuySizeDollars.Hundred);
+        // proxy.buyMeme{ value: 20 ether }(MemeBuySizeDollars.Ten);
```
Comment (hiddentao) [inline test/MemeTest.sol:134]: Looks like we need a loop here

Comment (kevin-fruitful) [reply test/MemeTest.sol:134]: will do after merge

File: src/shared/ReentrancyGuard.sol (hunk 7 above)
Comment (hiddentao) [reply src/shared/ReentrancyGuard.sol:36]: Since these values are only used in this file I think it's ok.

File: test/MemeTest.sol (hunk 8 above)
Comment (kevin-fruitful) [reply test/MemeTest.sol:92]: have a test for that - testPurchaseMemeRoutingEthToTreasury, but will make it more robust

File: test/MemeTest.sol (hunk 4 above)
Comment (kevin-fruitful) [reply test/MemeTest.sol]: done
